client/asset/eth/rpcclient.go

```diff
@@ -8,213 +8,320 @@ package eth
+		return nil, err
+	}
+
+	creds, err := nodeCredentials(node)
+	if err != nil {
+		return nil, err
+	}
+
+	return &nodeClient{
+		chainID:     big.NewInt(chainIDs[net]),
+		node:        node,
+		leth:        leth,
+		creds:       creds,
+		net:         net,
+		log:         log,
+		contractors: make(map[uint32]dexeth.Contractor),
+	}, nil
+}
+
+func (n *nodeClient) address() common.Address {
+	return n.creds.addr
 }
 
 // connect connects to a node. It then wraps ethclient's client and
 // bundles commands in a form we can easily use.
-func (c *rpcclient) connect(ctx context.Context, node *node.Node, contractAddr *common.Address) error {
-	client, err := node.Attach()
+func (n *nodeClient) connect(ctx context.Context) error {
+	client, err := n.node.Attach()
 	if err != nil {
 		return fmt.Errorf("unable to dial rpc: %v", err)
 	}
-	c.c = client
-	c.ec = ethclient.NewClient(client)
-	c.n = node
-	c.es, err = swap.NewETHSwap(*contractAddr, c.ec)
-	if err != nil {
-		return fmt.Errorf("unable to find swap contract: %v", err)
+	n.ec = ethclient.NewClient(client)
+	n.p2pSrv = n.node.Server()
+	if n.p2pSrv == nil {
+		return fmt.Errorf("no *p2p.Server")
 	}
+
+	n.es, _ = dexeth.NewETHSwap(dexeth.ContractAddresses[0][dex.Simnet], n.ec)
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	for ver, constructor := range dexeth.ContractorContstructors {
+		if n.contractors[ver], err = constructor(n.net, n.creds.addr, n.ec); err != nil {
+			return fmt.Errorf("error constructor version %d Contractor: %v", ver, err)
+		}
+	}
+
 	return nil
 }
 
 // shutdown shuts down the client.
-func (c *rpcclient) shutdown() {
-	if c.ec != nil {
-		// this will also close c.c
-		c.ec.Close()
+func (n *nodeClient) shutdown() {
+	if n.ec != nil {
+		// this will also close the underlying rpc.Client.
+		n.ec.Close()
+	}
+	if n.node != nil {
+		n.node.Close()
+		n.node.Wait()
 	}
 }
 
+// withContractor runs the provided function with the versioned Contractor.
+func (n *nodeClient) withContractor(ver uint32, f func(dexeth.Contractor) error) error {
+	contractor, found := n.contractors[ver]
+	if !found {
+		return fmt.Errorf("no version %d contractor", ver)
+	}
+	return f(contractor)
+}
+
 // bestBlockHash gets the best block's hash at the time of calling.
-func (c *rpcclient) bestBlockHash(ctx context.Context) (common.Hash, error) {
-	header, err := c.bestHeader(ctx)
+func (n *nodeClient) bestBlockHash(ctx context.Context) (common.Hash, error) {
+	header, err := n.bestHeader(ctx)
 	if err != nil {
 		return common.Hash{}, err
 	}
 	return header.Hash(), nil
 }
 
 // bestHeader gets the best header at the time of calling.
-func (c *rpcclient) bestHeader(ctx context.Context) (*types.Header, error) {
-	bn, err := c.ec.BlockNumber(ctx)
-	if err != nil {
-		return nil, err
-	}
-	header, err := c.ec.HeaderByNumber(ctx, big.NewInt(int64(bn)))
-	if err != nil {
-		return nil, err
-	}
-	return header, nil
+func (n *nodeClient) bestHeader(ctx context.Context) (*types.Header, error) {
```

Comment:
Plans for `Header` return, or remove?

Reply:
Hmm, I think I meant to comment on `pendingState`.  One of these had their callers always discarding the header return, but clearly it's not `bestHeader`